rectools/metrics/debias.py

```diff
@@ -0,0 +1,166 @@
+    iqr_coef : float, default 1.5
+        Coefficient for defining as the maximum value inside the border.
+    random_state : float, default None
+        Pseudorandom number generator state to control the down-sampling.
+    """
+
+    iqr_coef: float = attr.ib(default=1.5)
+    random_state: tp.Optional[int] = attr.ib(default=None)
+
+
+@attr.s
+class DebiasableMetrikAtK(MetricAtK):
+    """
+    Classification metric base class.
+
+    Warning: This class should not be used directly.
+    Use derived classes instead.
+
+    Parameters
+    ----------
+    k : int
+        Number of items at the top of recommendations list that will be used to calculate metric.
+    debias_config : DebiasConfig, default None
+        Config with debias method parameters (iqr_coef, random_state).
+    """
+
+    debias_config: DebiasConfig = attr.ib(default=None)
+
+    def _check_debias(self, is_debiased: bool, name_data: str) -> None:
+        if not is_debiased and self.debias_config is not None:
+            raise ValueError(
+                "You have specified `debias_config` for metric "
+                f"but `{name_data}` is not assumed to be de-biased. "
+                f"Please make de-biasing for `{name_data}` "
+                "and specify `is_debiased` as `True` "
+                "or otherwise use `calc` and `calc_per_user` methods for auto de-biasing."
+            )
+
+    def make_debias(self, interactions: pd.DataFrame) -> pd.DataFrame:
+        """
+        Downsample the size of interactions, excluding some interactions with popular items.
+
+        Algorithm:
+
+            1. Calculate item "popularity"
+            (here: number of unique users that had interaction with the item) distribution from interactions;
+            2. Find first (Q1) and third (Q3) quartiles in items "popularity" distribution;
+            3. Calculate IQR = Q3 - Q1;
+            4. Calculate maximum value inside by formula: Q3 + iqr_coef * IQR;
+            5. Down-sample for all exceeding items in interactions,
+            randomly keeping the maximum group of users to a size not exceeding
+            maximum value inside
+
+        Parameters
+        ----------
+        interactions : pd.DataFrame
+            Table with previous user-item interactions,
+            with columns `Columns.User`, `Columns.Item`.
+        iqr_coef : float, default 1.5
+            Coefficient for defining as the maximum value inside the border.
+        random_state: float, default 32
+            Pseudorandom number generator state to control the down-sampling.
+
+        Returns
+        -------
+        pd.DataFrame
+            Downsampling interactions.
+        """
+        if len(interactions) == 0:
+            return interactions
+
+        interactions_for_debiasing = interactions.copy()
+
+        num_users_interacted_with_item = interactions_for_debiasing.drop_duplicates(subset=Columns.UserItem)[
+            Columns.Item
+        ].value_counts()
```

Comment:
A bit shorter, faster and more memory efficient
```suggestion
        num_users_interacted_with_item = interactions_for_debiasing.groupby(Columns.Item, sort=False)[Columns.User].nunique()
```